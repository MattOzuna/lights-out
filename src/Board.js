import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows=3, ncols=3, chanceLightStartsOn=.5 }) {
  const [board, setBoard] = useState(createBoard());
  /** create a board nrows high/ncols wide, each cell randomly lit or unlit 
   *
   * Is there a better way to do this??
  */
  function createBoard() {
    let initialBoard = [];    
    for(let i = 0; i < nrows; i++){
      let col = [];
      for (let k = 0; k < ncols; k++){
        if(Math.random() < chanceLightStartsOn) col.push(true)
        else col.push(false)
      }
      initialBoard.push(col)
    }
    return initialBoard;
  }

  /**Checks board state if all lights are off
   * returns boolean
   */
  function hasWon() {
    return board.reduce((acc, row) => {
      const rowVal = row.reduce((acc, val) => {
        if(!val && acc) return true
        return false
      },true)
      if(rowVal && acc) return true
      return false
    }, true)
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      // TODO: Make a (deep) copy of the oldBoard
      let boardCopy = board.map(row => [...row])

      // TODO: in the copy, flip this cell and the cells around it
      flipCell(y,x, boardCopy)
      flipCell(y-1,x, boardCopy)
      flipCell(y+1,x, boardCopy)
      flipCell(y,x-1, boardCopy)
      flipCell(y,x+1, boardCopy)

      // TODO: return the copy
      return boardCopy
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  if(hasWon()){
    return(
      <div className="Board">
        <h1>You Win!</h1>
      </div>
    )
  }

  // make table board
  return (
    <table className="Board">
      <tbody>
      {board.map((row, y) => (
          <tr key={y}>
            {row.map((val, x) => (
              <Cell 
                isLit={val} 
                flipCellsAroundMe={()=>flipCellsAround(`${y}-${x}`)}
                id={`${y}${x}`}
                key={x}
              />
            ))}          
          </tr>
        ))}
      </tbody>
    </table>
    )
}
export default Board;
