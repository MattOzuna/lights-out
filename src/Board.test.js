import React from "react";
import {render, asFragment, fireEvent} from "@testing-library/react"
import Board from "./Board.js"

it('should render properly', ()=>{
    render(
        <Board
            nrows={3}
            ncols={3}
            chanceLightStartsOn={0}
        />
    )
})

it('should match snapshot of non winning game', ()=>{
    const {asFragment} = render(
        <Board
            nrows={3}
            ncols={3}
            chanceLightStartsOn={1}
        />
    )
    expect(asFragment()).toMatchSnapshot();
})

it('should match snapshot of winning game', ()=>{
    const {asFragment} = render(
        <Board
            nrows={3}
            ncols={3}
            chanceLightStartsOn={0}
        />
    )
    expect(asFragment()).toMatchSnapshot();
})

it('should light correct Cells when clicked', ()=>{
    //Create a board with all cells lit
    const {container} = render(
        <Board
            nrows={3}
            ncols={3}
            chanceLightStartsOn={1}
        />
    );0

    //all cells should be lit
    expect(container.querySelector(`[id="00"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="01"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="02"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="10"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="11"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="12"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="20"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="21"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="22"]`)).toHaveClass('Cell-lit')


    //click the middle cell with coord 1-1 matching its Id
    const testCell = container.querySelector(`[id="11"]`)
    fireEvent.click(testCell)
    
    //the middle cella and all surrounding cells should not have the class cell-lit
    //the corner cells should still be lit
    expect(container.querySelector(`[id="11"]`)).not.toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="12"]`)).not.toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="10"]`)).not.toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="01"]`)).not.toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="21"]`)).not.toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="00"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="02"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="20"]`)).toHaveClass('Cell-lit')
    expect(container.querySelector(`[id="22"]`)).toHaveClass('Cell-lit')
    
})
    



