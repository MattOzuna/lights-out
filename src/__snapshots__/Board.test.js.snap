// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot of non winning game 1`] = `
<DocumentFragment>
  <table>
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          id="00"
        />
        <td
          class="Cell Cell-lit"
          id="01"
        />
        <td
          class="Cell Cell-lit"
          id="02"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          id="10"
        />
        <td
          class="Cell Cell-lit"
          id="11"
        />
        <td
          class="Cell Cell-lit"
          id="12"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          id="20"
        />
        <td
          class="Cell Cell-lit"
          id="21"
        />
        <td
          class="Cell Cell-lit"
          id="22"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`should match snapshot of winning game 1`] = `
<DocumentFragment>
  <div>
    <h1>
      You Win!
    </h1>
  </div>
</DocumentFragment>
`;
